src/Engine/ProtoCore/Utils/CompilerUtils.cs

```diff
@@ -280,16 +281,19 @@ public static bool PreCompileCodeBlock(Core core, ref ParseParam parseParams, ID
             List<AssociativeNode> astNodes;
             List<AssociativeNode> comments;
             ParseUserCode(core, parseParams.OriginalCode, postfixGuid, out astNodes, out comments);
-
-            // Catch the syntax errors and errors for unsupported 
-            // language constructs thrown by compile expression
-            if (core.BuildStatus.ErrorCount > 0)
+            parseParams.AppendErrors(core.BuildStatus.Errors);
+            if (parseParams.Errors.Count() > 0)
             {
-                parseParams.AppendErrors(core.BuildStatus.Errors);
-                parseParams.AppendWarnings(core.BuildStatus.Warnings);
                 return false;
             }
-            parseParams.AppendParsedNodes(astNodes);
+
+            // Catch the syntax errors and errors for unsupported 
+            // language constructs thrown by compile expression
+            parseParams.AppendWarnings(core.BuildStatus.Warnings);
+            var warnings = Check(astNodes);
```

Comment:
I suggest a more descriptive name for `Check` method, maybe something like `CheckForImmutableVariables`?

Reply:
It actually checks immutable variables and recursive reference now. In the future we may add more checking to it.